File: containers/splunk/splunk__connectors/sql/python/src/handler.py
import asyncio
import json
import sqlite3
from collections import defaultdict
from contextlib import AsyncExitStack
from datetime import datetime
from typing import Dict, List, Optional, Tuple, Union
from urllib.parse import urlparse, parse_qs
import oracledb
import psycopg2
from psycopg2.extras import LogicalReplicationConnection
import pyodbc
import pymysql
from prometheus_client import Counter, Gauge, Histogram
import structlog
from tenacity import retry, stop_after_attempt, wait_exponential
from schema import Schema, Schemas
from secrets import token_hex

logger = structlog.get_logger()

EVENTS_PROCESSED = Counter("sql_connector_events_processed_total", "Total events processed per table", ["table"])
ERRORS_TOTAL = Counter("sql_connector_errors_total", "Total errors per table", ["table"])
QUERY_LATENCY = Histogram("sql_connector_query_latency_seconds", "Query latency per table", ["table"])
ACTIVE_CONNECTIONS = Gauge("sql_connector_active_connections", "Active database connections")

class HandlerError(Exception):
    pass

class Position:
    def __init__(self, kind: str, value: Union[datetime, str, int]):
        self.kind = kind
        self.value = value

    def to_dict(self) -> Dict:
        return {"kind": self.kind, "value": str(self.value) if isinstance(self.value, datetime) else self.value}

    @classmethod
    def from_dict(cls, data: Dict) -> "Position":
        kind = data["kind"]
        value = data["value"]
        if kind == "Timestamp":
            value = datetime.fromisoformat(value)
        elif kind == "Id":
            value = value  # UUID as string
        elif kind == "Lsn":
            value = int(value)
        return cls(kind, value)

class SqlHandler:
    def __init__(
        self,
        db_type: str,
        conn_str: str,
        sender_queue: asyncio.Queue,
        schemas: Schemas,
        batch_size: int,
        poll_interval: float,
        state_path: str,
        cdc_enabled: bool,
        max_connections_per_table: int,
        sqlcipher_key: str = token_hex(16),
    ):
        if not (1 <= batch_size <= 1000):
            raise HandlerError("batch_size must be between 1 and 1000")
        if not (1 <= poll_interval <= 3600):
            raise HandlerError("poll_interval must be between 1 and 3600 seconds")
        if not (1 <= max_connections_per_table <= 50):
            raise HandlerError("max_connections_per_table must be between 1 and 50")

        self.db_type = db_type.lower()
        self.conn_str = conn_str
        self.sender_queue = sender_queue
        self.schemas = schemas
        self.batch_size = batch_size
        self.poll_interval = poll_interval
        self.state_path = state_path
        self.cdc_enabled = cdc_enabled
        self.max_connections_per_table = max_connections_per_table
        self.sqlcipher_key = sqlcipher_key
        self.pools: Dict[str, List] = defaultdict(list)
        self.last_positions: Dict[str, Position] = self._load_state()
        self._validate_db_type()

    def _validate_db_type(self):
        valid_types = {"postgres", "mysql", "mssql", "sqlite", "oracle"}
        if self.db_type not in valid_types:
            raise HandlerError(f"Unsupported database type: {self.db_type}")

    def _load_state(self) -> Dict[str, Position]:
        try:
            with sqlite3.connect(f"file:{self.state_path}?mode=rw", uri=True) as conn:
                conn.execute(f"PRAGMA key='{self.sqlcipher_key}'")
                conn.execute("CREATE TABLE IF NOT EXISTS positions (table_name TEXT PRIMARY KEY, position TEXT)")
                cursor = conn.cursor()
                cursor.execute("SELECT table_name, position FROM positions")
                return {row[0]: Position.from_dict(json.loads(row[1])) for row in cursor.fetchall()}
        except sqlite3.Error as e:
            logger.error(f"State load error: {e}", exc_info=True)
            return {}

    def _save_state(self, table: str, position: Position):
        try:
            with sqlite3.connect(f"file:{self.state_path}?mode=rw", uri=True) as conn:
                conn.execute(f"PRAGMA key='{self.sqlcipher_key}'")
                conn.execute("INSERT OR REPLACE INTO positions (table_name, position) VALUES (?, ?)",
                             (table, json.dumps(position.to_dict())))
                conn.commit()
        except sqlite3.Error as e:
            logger.error(f"State save error: {e}", exc_info=True)

    async def start(self):
        if self.cdc_enabled and self.db_type == "postgres":
            await self._start_cdc()
        else:
            await self._start_polling()

    @retry(stop=stop_after_attempt(5), wait=wait_exponential(multiplier=1, min=0.1, max=10))
    async def _get_connection(self, table: Optional[str] = None):
        async with AsyncExitStack() as stack:
            if table and table in self.pools and len(self.pools[table]) < self.max_connections_per_table:
                ACTIVE_CONNECTIONS.set(sum(len(pool) for pool in self.pools.values()))
                return self.pools[table][-1]

            if self.db_type == "postgres":
                conn = stack.enter_context(psycopg2.connect(self.conn_str))
            elif self.db_type == "mysql":
                conn = stack.enter_context(pymysql.connect(**self._parse_mysql_conn_str()))
            elif self.db_type == "mssql":
                conn = stack.enter_context(pyodbc.connect(self.conn_str))
            elif self.db_type == "sqlite":
                conn = stack.enter_context(sqlite3.connect(self.conn_str))
            elif self.db_type == "oracle":
                conn = stack.enter_context(oracledb.connect(self.conn_str))
            else:
                raise HandlerError("Unknown database type")

            if table:
                if len(self.pools[table]) >= self.max_connections_per_table:
                    self.pools[table].pop(0).close()
                self.pools[table].append(conn)
            ACTIVE_CONNECTIONS.set(sum(len(pool) for pool in self.pools.values()))
            return conn

    def _parse_mysql_conn_str(self) -> Dict:
        parsed = urlparse(self.conn_str)
        if parsed.scheme != "mysql":
            raise HandlerError("Invalid MySQL connection string")
        params = parse_qs(parsed.query)
        return {
            "user": parsed.username,
            "password": parsed.password,
            "host": parsed.hostname,
            "port": parsed.port or 3306,
            "database": parsed.path.lstrip("/") or params.get("db", [""])[0],
        }

    async def _discover_tables(self) -> List[str]:
        logger.info("Discovering tables")
        async with AsyncExitStack() as stack:
            conn = await self._get_connection(None)
            stack.push(conn)
            try:
                start = datetime.now()
                cursor = conn.cursor()
                if self.db_type == "postgres":
                    cursor.execute("SELECT table_name FROM information_schema.tables WHERE table_schema = 'public' AND (table_type = 'BASE TABLE' OR table_type = 'VIEW')")
                elif self.db_type == "mysql":
                    cursor.execute("SHOW TABLES")
                elif self.db_type == "mssql":
                    cursor.execute("SELECT TABLE_NAME FROM INFORMATION_SCHEMA.TABLES WHERE TABLE_TYPE = 'BASE TABLE' OR TABLE_TYPE = 'VIEW'")
                elif self.db_type == "sqlite":
                    cursor.execute("SELECT name FROM sqlite_master WHERE type='table' OR type='view'")
                elif self.db_type == "oracle":
                    cursor.execute("SELECT TABLE_NAME FROM USER_TABLES UNION SELECT VIEW_NAME AS TABLE_NAME FROM USER_VIEWS")

                tables = [row[0] for row in cursor.fetchall()]
                QUERY_LATENCY.labels("discover").observe((datetime.now() - start).total_seconds())
                logger.debug(f"Discovered tables: {tables}")
                return tables
            except Exception as e:
                logger.error(f"Table discovery error: {e}", exc_info=True)
                raise HandlerError(f"Failed to discover tables: {e}")

    async def _start_polling(self):
        while True:
            try:
                tables = await self._discover_tables()
                for table in tables:
                    try:
                        await self._process_table(table)
                    except Exception as e:
                        ERRORS_TOTAL.labels(table).inc()
                        logger.error(f"Error processing table {table}: {e}", exc_info=True)
            except Exception as e:
                ERRORS_TOTAL.labels("discover").inc()
                logger.error(f"Error discovering tables: {e}", exc_info=True)
            await asyncio.sleep(self.poll_interval)

    async def _start_cdc(self):
        if self.db_type != "postgres":
            raise HandlerError("CDC only supported for PostgreSQL")

        async with AsyncExitStack() as stack:
            conn = stack.enter_context(psycopg2.connect(self.conn_str, connection_factory=LogicalReplicationConnection))
            cursor = stack.enter_context(conn.cursor())
            try:
                cursor.execute("CREATE PUBLICATION sql_connector_pub FOR ALL TABLES")
                cursor.execute("CREATE_REPLICATION_SLOT sql_connector_slot LOGICAL pgoutput")
                cursor.start_replication(slot_name="sql_connector_slot", decode=True)

                cdc_queue = asyncio.Queue(maxsize=self.batch_size)

                async def process_messages():
                    while True:
                        batch = []
                        start = datetime.now()
                        for _ in range(self.batch_size):
                            try:
                                msg = await asyncio.wait_for(cdc_queue.get(), timeout=1.0)
                                if msg.data_type == "INSERT":
                                    table = msg.relation_name
                                    if schema := self.schemas.get_schema(table):
                                        event = {col: val for col, val in zip(msg.column_names, msg.column_values)}
                                        transformed = self._transform_to_cim(event, schema)
                                        batch.append(transformed)
                                        EVENTS_PROCESSED.labels(table).inc()
                                        if lsn := msg.lsn:
                                            self.last_positions[table] = Position("Lsn", lsn)
                                            self._save_state(table, self.last_positions[table])
                                msg.ack()
                            except asyncio.TimeoutError:
                                break
                        if batch:
                            await self.sender_queue.put((batch, table, len(batch)))
                        QUERY_LATENCY.labels("cdc").observe((datetime.now() - start).total_seconds())

                def consume_stream():
                    def process_msg(msg):
                        asyncio.create_task(cdc_queue.put(msg))
                    cursor.consume_stream(process_msg)

                loop = asyncio.get_running_loop()
                await loop.run_in_executor(None, consume_stream)
            except Exception as e:
                logger.error(f"CDC error: {e}", exc_info=True)
                raise HandlerError(f"CDC failed: {e}")

    async def _process_table(self, table: str):
        schema = self.schemas.get_schema(table)
        if not schema:
            logger.warning(f"No schema found for table {table}")
            return

        logger.info(f"Processing table: {table}")
        async with AsyncExitStack() as stack:
            conn = await self._get_connection(table)
            stack.push(conn)
            try:
                start = datetime.now()
                query = f'SELECT * FROM "{table}" WHERE 1=1'
                pos = self.last_positions.get(table)
                order_field = "1"
                filter_clause = ""

                if schema.timestamp_field and pos and pos.kind == "Timestamp":
                    order_field = schema.timestamp_field
                    filter_clause = f' AND "{schema.timestamp_field}" > \'{pos.value}\''
                elif schema.id_field and pos and pos.kind == "Id":
                    order_field = schema.id_field
                    filter_clause = f' AND "{schema.id_field}" > \'{pos.value}\''
                elif schema.timestamp_field:
                    order_field = schema.timestamp_field
                elif schema.id_field:
                    order_field = schema.id_field
                else:
                    logger.warning(f"No incremental field for table {table}, full scan")

                query += filter_clause + f' ORDER BY "{order_field}" ASC LIMIT {self.batch_size}'
                logger.debug(f"Executing query: {query}")

                cursor = conn.cursor()
                cursor.execute(query)
                rows = cursor.fetchall()
                columns = [desc[0] for desc in cursor.description]
                QUERY_LATENCY.labels(table).observe((datetime.now() - start).total_seconds())

                batch = []
                new_pos = pos
                for row in rows:
                    event = dict(zip(columns, row))
                    transformed = self._transform_to_cim(event, schema)
                    batch.append(transformed)

                    if schema.timestamp_field and (ts := event.get(schema.timestamp_field)):
                        new_pos = Position("Timestamp", ts)
                    elif schema.id_field and (id_ := event.get(schema.id_field)):
                        new_pos = Position("Id", id_)

                    EVENTS_PROCESSED.labels(table).inc()

                if new_pos:
                    self.last_positions[table] = new_pos
                    self._save_state(table, new_pos)

                if batch:
                    logger.info(f"Sending batch of {len(batch)} events for table {table}")
                    await self.sender_queue.put((batch, table, len(batch)))
            except Exception as e:
                ERRORS_TOTAL.labels(table).inc()
                logger.error(f"Table processing error: {e}", exc_info=True)
                raise HandlerError(f"Failed to process table {table}: {e}")

    def _transform_to_cim(self, event: Dict, schema: Schema) -> Dict:
        cim = {
            "time": event.get(schema.mappings.cim.get("time", "timestamp"), datetime.utcnow().timestamp()),
            "vendor_product": "SQL_Connector",
            "schema": schema.name,
        }

        for key, value in schema.mappings.cim.items():
            if key != "sourcetype":
                if value.startswith('"') and value.endswith('"'):
                    cim[key] = value.strip('"')
                else:
                    cim[key] = event.get(value, "")

        return {
            "event": cim,
            "sourcetype": schema.mappings.cim.get("sourcetype", f"sql:{schema.name}")
        }

File: containers/splunk/splunk__connectors/sql/rust/src/handler.rs
use std::collections::HashMap;
use std::path::Path;
use std::sync::Arc;
use std::time::{Duration, Instant};

use chrono::{DateTime, Utc};
use postgres_protocol::message::backend::ReplicationMessage;
use postgres_types::{FromSql, Type};
use prometheus::{IntCounterVec, IntGauge, HistogramVec};
use serde::{Deserialize, Serialize};
use serde_json::{json, Value};
use sled::Db;
use sqlx::{any::AnyRow, AnyPool, AnyPoolOptions, Connection, Executor, Pool, Row};
use thiserror::Error;
use tokio::sync::mpsc;
use tokio_retry::{strategy::ExponentialBackoff, Retry};
use tracing::{debug, error, info, warn};
use uuid::Uuid;
use warp::Filter;

use crate::schema::{Schema, Schemas};

lazy_static::lazy_static! {
    static ref EVENTS_PROCESSED: IntCounterVec = prometheus::register_int_counter_vec!(
        "sql_connector_events_processed_total",
        "Total number of events processed per table",
        &["table"]
    ).unwrap();
    static ref ERRORS_TOTAL: IntCounterVec = prometheus::register_int_counter_vec!(
        "sql_connector_errors_total",
        "Total number of errors per table",
        &["table"]
    ).unwrap();
    static ref QUERY_LATENCY: HistogramVec = prometheus::register_histogram_vec!(
        "sql_connector_query_latency_seconds",
        "Query latency per table",
        &["table"]
    ).unwrap();
    static ref ACTIVE_CONNECTIONS: IntGauge = prometheus::register_int_gauge!(
        "sql_connector_active_connections",
        "Number of active database connections"
    ).unwrap();
}

#[derive(Clone)]
pub enum DbType {
    Postgres,
    MySql,
    MsSql,
    SQLite,
    Oracle,
}

#[derive(Error, Debug)]
pub enum HandlerError {
    #[error("Database error: {0}")]
    Db(#[from] sqlx::Error),
    #[error("Serialization error: {0}")]
    Serde(#[from] serde_json::Error),
    #[error("State storage error: {0}")]
    Sled(#[from] sled::Error),
    #[error("Send error: {0}")]
    Send(#[from] mpsc::error::SendError<(Vec<Value>, String, u64)>),
    #[error("CDC error: {0}")]
    Cdc(String),
    #[error("Other error: {0}")]
    Other(#[from] Box<dyn std::error::Error + Send + Sync>),
}

#[derive(Serialize, Deserialize, Clone, Copy)]
pub enum Position {
    Timestamp(DateTime<Utc>),
    Id(Uuid),
    Lsn(u64), // For PostgreSQL CDC
}

pub struct SqlHandler {
    db_type: DbType,
    conn_str: String,
    tx: mpsc::Sender<(Vec<Value>, String, u64)>,
    schemas: Arc<Schemas>,
    batch_size: usize,
    last_positions: HashMap<String, Position>,
    poll_interval: Duration,
    state_db: Db,
    cdc_enabled: bool,
    max_connections_per_table: u32,
    pools: HashMap<String, Pool<sqlx::Any>>,
}

impl SqlHandler {
    pub fn new(
        db_type: DbType,
        conn_str: String,
        tx: mpsc::Sender<(Vec<Value>, String, u64)>,
        schemas: Arc<Schemas>,
        batch_size: usize,
        poll_interval: Duration,
        state_path: &str,
        cdc_enabled: bool,
        max_connections_per_table: u32,
    ) -> Result<Self, HandlerError> {
        let state_db = sled::open(state_path)?;
        let mut last_positions = HashMap::new();
        for res in state_db.iter() {
            let (key, value) = res?;
            let table = String::from_utf8(key.to_vec()).unwrap_or_default();
            let pos: Position = bincode::deserialize(&value)?;
            last_positions.insert(table, pos);
        }
        Ok(SqlHandler {
            db_type,
            conn_str,
            tx,
            schemas,
            batch_size,
            last_positions,
            poll_interval,
            state_db,
            cdc_enabled,
            max_connections_per_table,
            pools: HashMap::new(),
        })
    }

    pub async fn start(&mut self) -> Result<(), HandlerError> {
        // Start metrics server
        tokio::spawn(async move {
            let metrics = warp::path("metrics")
                .and(warp::get())
                .map(|| {
                    let metric_families = prometheus::gather();
                    warp::reply::with_header(
                        prometheus::TextEncoder::new().encode_to_string(&metric_families).unwrap(),
                        "content-type",
                        "text/plain; version=0.0.4",
                    )
                });
            warp::serve(metrics).run(([0, 0, 0, 0], 9000)).await;
        });

        if self.cdc_enabled && matches!(self.db_type, DbType::Postgres) {
            self.start_cdc().await?;
        } else {
            self.start_polling().await?;
        }
        Ok(())
    }

    async fn start_polling(&mut self) -> Result<(), HandlerError> {
        loop {
            match self.discover_tables().await {
                Ok(tables) => {
                    for table in tables {
                        if let Err(e) = self.process_table(&table).await {
                            ERRORS_TOTAL.with_label_values(&[&table]).inc();
                            error!("Error processing table {}: {}", table, e);
                        }
                    }
                }
                Err(e) => {
                    ERRORS_TOTAL.with_label_values(&["discover"]).inc();
                    error!("Error discovering tables: {}", e);
                }
            }
            tokio::time::sleep(self.poll_interval).await;
        }
    }

    async fn start_cdc(&mut self) -> Result<(), HandlerError> {
        if !matches!(self.db_type, DbType::Postgres) {
            return Err(HandlerError::Cdc("CDC only supported for PostgreSQL".into()));
        }

        let pool = self.get_pool(None).await?;
        let mut conn = pool.acquire().await?;
        sqlx::query("CREATE PUBLICATION sql_connector_pub FOR ALL TABLES").execute(&mut *conn).await?;
        let slot_name = "sql_connector_slot";
        sqlx::query(&format!("CREATE_REPLICATION_SLOT {} LOGICAL pgoutput", slot_name)).execute(&mut *conn).await?;

        let mut stream = sqlx::postgres::PgCopyIn::new(&mut *conn, &format!("START_REPLICATION SLOT {} LOGICAL 0/0", slot_name)).await?;
        while let Some(replication_message) = stream.next().await {
            match replication_message {
                ReplicationMessage::XLogData(xlog_data) => {
                    if let Some(change) = xlog_data.data().as_change() {
                        self.process_cdc_change(change).await?;
                    }
                }
                _ => continue,
            }
        }
        Ok(())
    }

    async fn process_cdc_change(&mut self, change: postgres_protocol::message::backend::LogicalReplicationChange) -> Result<(), HandlerError> {
        match change {
            postgres_protocol::message::backend::LogicalReplicationChange::Insert { relation, tuple } => {
                let table = relation.name().to_string();
                if let Some(schema) = self.schemas.get_schema(&table) {
                    let event: HashMap<String, String> = relation
                        .columns()
                        .iter()
                        .zip(tuple.values())
                        .filter_map(|(col, val)| {
                            val.as_ref().map(|v| (col.name().to_string(), v.to_string()))
                        })
                        .collect();
                    let transformed = self.transform_to_cim(&event, schema);
                    self.tx.send((vec![transformed], table.clone(), 1)).await?;
                    EVENTS_PROCESSED.with_label_values(&[&table]).inc();

                    if let Some(lsn) = change.lsn() {
                        self.last_positions.insert(table.clone(), Position::Lsn(lsn));
                        self.state_db.insert(table.as_bytes(), &bincode::serialize(&Position::Lsn(lsn))?)?;
                        self.state_db.flush()?;
                    }
                }
            }
            _ => {} // Handle updates/deletes if needed
        }
        Ok(())
    }

    async fn discover_tables(&self) -> Result<Vec<String>, HandlerError> {
        info!("Discovering tables");
        let pool = self.get_pool(None).await?;
        let query = match self.db_type {
            DbType::Postgres => "SELECT table_name FROM information_schema.tables WHERE table_schema = 'public' AND (table_type = 'BASE TABLE' OR table_type = 'VIEW')",
            DbType::MySql => "SHOW TABLES",
            DbType::MsSql => "SELECT TABLE_NAME FROM INFORMATION_SCHEMA.TABLES WHERE TABLE_TYPE = 'BASE TABLE' OR TABLE_TYPE = 'VIEW'",
            DbType::SQLite => "SELECT name FROM sqlite_master WHERE type='table' OR type='view'",
            DbType::Oracle => "SELECT TABLE_NAME FROM USER_TABLES UNION SELECT VIEW_NAME AS TABLE_NAME FROM USER_VIEWS",
        };
        let start = Instant::now();
        let rows = sqlx::query(query).fetch_all(&pool).await?;
        let latency = start.elapsed().as_secs_f64();
        QUERY_LATENCY.with_label_values(&["discover"]).observe(latency);
        let tables: Vec<String> = rows.iter().map(|row: &AnyRow| row.get(0)).collect();
        debug!("Discovered tables: {:?}", tables);
        Ok(tables)
    }

    async fn get_pool(&mut self, table: Option<&str>) -> Result<Pool<sqlx::Any>, HandlerError> {
        if let Some(table) = table {
            if let Some(pool) = self.pools.get(table) {
                ACTIVE_CONNECTIONS.set(pool.status().size as i64);
                return Ok(pool.clone());
            }
        }
        let retry_strategy = ExponentialBackoff::from_millis(100).max_delay(Duration::from_secs(10)).take(5);
        let pool = Retry::retry_async(retry_strategy, || async {
            info!("Creating connection pool for {}", table.unwrap_or("global"));
            AnyPoolOptions::new()
                .max_connections(self.max_connections_per_table)
                .connect(&self.conn_str)
                .await
                .map_err(HandlerError::from)
        })
        .await?;
        ACTIVE_CONNECTIONS.set(pool.status().size as i64);
        if let Some(table) = table {
            self.pools.insert(table.to_string(), pool.clone());
        }
        Ok(pool)
    }

    async fn process_table(&mut self, table: &str) -> Result<(), HandlerError> {
        if let Some(schema) = self.schemas.get_schema(table) {
            info!("Processing table: {}", table);
            let pool = self.get_pool(Some(table)).await?;
            let pos = self.last_positions.get(table).cloned();

            let mut query = format!("SELECT * FROM \"{}\" WHERE 1=1", table);
            let (order_field, filter) = match (&schema.timestamp_field, &schema.id_field, pos) {
                (Some(ts), _, Some(Position::Timestamp(last_ts))) => (ts.clone(), format!(" AND \"{}\" > '{}'", ts, last_ts)),
                (_, Some(id), Some(Position::Id(last_id))) => (id.clone(), format!(" AND \"{}\" > '{}'", id, last_id)),
                (Some(ts), _, _) => (ts.clone(), String::new()),
                (_, Some(id), _) => (id.clone(), String::new()),
                _ => {
                    warn!("No incremental field for table {}, full scan", table);
                    ("1".to_string(), String::new())
                },
            };
            query.push_str(&filter);
            query.push_str(&format!(" ORDER BY \"{}\" ASC LIMIT {}", order_field, self.batch_size));

            debug!("Executing query: {}", query);
            let start = Instant::now();
            let mut rows = sqlx::query(&query).fetch(&pool);
            let mut batch = Vec::new();
            let mut new_pos = pos;
            let mut event_count = 0;

            while let Some(row) = rows.try_next().await? {
                let event: HashMap<String, String> = (0..row.len())
                    .filter_map(|i| {
                        let name = row.column(i).name().to_string();
                        let value: Option<String> = row.try_get(i).ok();
                        value.map(|v| (name, v))
                    })
                    .collect();
                let transformed = self.transform_to_cim(&event, schema);
                batch.push(transformed);

                if let Some(ts_field) = &schema.timestamp_field {
                    if let Ok(ts) = row.try_get(ts_field) {
                        new_pos = Some(Position::Timestamp(ts));
                    }
                } else if let Some(id_field) = &schema.id_field {
                    if let Ok(id) = row.try_get(id_field) {
                        new_pos = Some(Position::Id(id));
                    }
                }

                event_count += 1;
                EVENTS_PROCESSED.with_label_values(&[table]).inc();
                if event_count >= self.batch_size {
                    break;
                }
            }

            QUERY_LATENCY.with_label_values(&[table]).observe(start.elapsed().as_secs_f64());

            if let Some(pos) = new_pos {
                self.last_positions.insert(table.to_string(), pos);
                self.state_db.insert(table.as_bytes(), &bincode::serialize(&pos)?)?;
                self.state_db.flush()?;
            }

            if !batch.is_empty() {
                info!("Sending batch of {} events for table {}", batch.len(), table);
                self.tx.send((batch, table.to_string(), event_count as u64)).await?;
            }
        } else {
            warn!("No schema found for table {}", table);
        }
        Ok(())
    }

    fn transform_to_cim(&self, event: &HashMap<String, String>, schema: &Schema) -> Value {
        let mut cim = json!({
            "time": event.get(schema.mappings.cim.get("time").unwrap_or(&"timestamp".to_string())).unwrap_or(&chrono::Utc::now().timestamp().to_string()),
            "vendor_product": "SQL_Connector",
            "schema": &schema.name
        });

        for (key, value) in &schema.mappings.cim {
            if key != "sourcetype" {
                let target_value = if value.starts_with('"') && value.ends_with('"') {
                    json!(value.trim_matches('"'))
                } else {
                    json!(event.get(value).unwrap_or_default())
                };
                cim[key] = target_value;
            }
        }

        json!({
            "event": cim,
            "sourcetype": schema.mappings.cim.get("sourcetype").map_or(format!("sql:{}", schema.name), |v| v.to_string())
        })
    }
}